src/QirRuntime/lib/Tracer/README.md

```diff
@@ -0,0 +1,177 @@
+ program can be also placed into a single layer L(0, infinity). Some gates might be considered to be very cheap and take
+ zero time to execute, those gates can be added to a layer even if they act on the same qubit another gate in this layer
+ is acting on and have to be executed sequentially within the layer.
+
+_Definition_: A ___global barrier___ is any operation that acts on _all_ currently allocated qubits. We'll provide means
+ for the clients to inject global barriers equivalent to an identity operator. This will allow the clients to enforce a
+ particular layering structure (because no later operation can sink below the barrier).
+
+The global barriers will be implemented as calls to `__quantum__qis__global_barrier` function. No additional support from
+ the compiler should be needed, as the user can define their own intrinsic to represent the barrier and map it to the above
+ runtime function via targets.qs file. The user can choose duration of a barrier which would affect start time of the
+ following layers but no operations will be added to a barrier, independent of its width.
+
+### The Resource Tracer's Layering Algorithm ###
+
+As the tracer is executing a sequential quantum program, it will compute a time function and corresponding layering
+ using the _conceptual_ algorithm, described below (aka "tetris algorithm"). The actual implementation of layering might
+ be done differently, as long as the resulting layering is the same as if running the conceptual algorithm.
+
+1. The tracer will have a setting for preferred layer duration: P.
+1. The first encountered operation of __non-zero__ duration N is added into layer L(0, max(P,N)). The value
+ for _conditional barrier_ is set to 0.
+1. When conditional callback is encountered, the layer L(t,N) of the measurement that produced the result the conditional
+ is dependent on, is looked up and the _conditional barrier_ is set to _t + N_. At the end of the conditional scope the
+ barrier is reset to 0. (Effectively, no operations, conditioned on the result of a measurement, can happen before or in
+ the same layer as the measurement, even if they don't involve the measured qubits.)
+1. Suppose, there are already layers L(0,N0), ... , L(k,Nk) and the operation being executed is a single-qubit _op_ of
+ duration __0__ (controlled and multi-qubit operations of duration 0 are treated the same as non-zero operations).
+ Starting at L(k, Nk) and scanning backwards to L(conditional barrier, Nb) find the _first_ layer that contains an
+ operation that acts on at least one of the qubits _op_ is acting on. Add _op_ into this layer. If no such layer found,
+ remember that the qubits have pending 0-duration _op_ and commit it to the layer that first uses any of the qubits in a
+ an operation of non-zero duration. At the end of the program commit all pending operations of duration zero into a new
+ layer.

```

Comment:
> that the **qubits have** pending 0-duration _op_ and **commit it** to the layer that first uses **any of the qubits** in **a** an operation of non-zero duration.

I can interpret this in two ways:
* that **this qubit has** a pending 0-duration operation and **add the operation** to the **later** layer that first uses **this qubit** in an operation of non-zero duration **(until then all the zero-duration operations on this qubit are memorized)**.
* that there is a pending (0-duration single-qubit) operation and **add the operation** to the **later** layer that first uses a non-zero-duration operation **on any qubit** **(until then \<I need to know more to word this part>)**.